source/common/stats/heap_stat_data.cc

```diff
@@ -204,13 +198,29 @@ class GaugeImpl : public Gauge, public MetricImpl {
 CounterSharedPtr HeapStatDataAllocator::makeCounter(StatName name,
                                                     absl::string_view tag_extracted_name,
                                                     const std::vector<Tag>& tags) {
-  return std::make_shared<CounterImpl>(alloc(name), *this, tag_extracted_name, tags);
+  Thread::LockGuard lock(mutex_);
+  auto iter = counters_.find(name);
+  if (iter != counters_.end()) {
+    return CounterSharedPtr(*iter);
+  }
+  Counter* counter =
```

Comment:
nit: prefer allocate directly into shared_ptr and then use get to insert into the map, same below.

Reply:
done